main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -249,15 +253,28 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
       else
         Used.empty + HiddenVarSym(sym, loc)
 
-    case Expression.Def(sym, _, _, _) => Used.of(sym)
+    case Expression.Def(sym, _, _, _) => {
+      if (env0.recursionContext.exists(r => r.defn == sym && r.applies >= r.params)) {
+        Used.of(sym, unconditionallyRecurses = true)
+      } else {
+        Used.of(sym, unconditionallyRecurses = false)
+      }
+    }
 
     case Expression.Eff(sym, _, _, _) => Used.empty
 
     case Expression.Hole(sym, _, _, _) => Used.of(sym)
 
     case Expression.Lambda(fparam, exp, _, _, _) =>
       // Extend the environment with the variable symbol.
-      val env1 = env0 + fparam.sym
+      val envExtended = env0 + fparam.sym
+
+      // Expand the recursion context parameters if immediately under a definition
```

Comment:
This is a bit of hack. Maybe we can use the type of the def, at the top-level, to simply count how many arguments are required? I think that should work and avoids this messy business of counting lambdas.

Could you try to see if it is possible to write a helper method:

```
def arity(defn: TypedAst.Def): Int
```
?

Reply:
Then we can kill addParam and stopBuilding.

Reply:
Yes, I think that would definitely help clean things up. Thanks!